### PR TITLE
Add dark theme support

### LINKED ISSUE #39: Syntax Highlight has wrong colors
image

Compare to GitHub: https://github.com/Microsoft/vscode-docs/blob/ext-docs/api/extension-guides/tree-view.md

image

### LINKED ISSUE #65: Dark theme
GitHub now has official support for 2 themes:

- Default dark
- Dark dimmed

Having at least one of them implemented in this extension (and used if active theme is dark) would be great.

### PR DESCRIPTION
**Pull request by SNDST00M**
Once they spoke of it in hushed tones... it is now time.
Closes #65.
Fixes #39.

### CODE REVIEW

**Review by mjbvz — COMMENTED** [review]
This looks great overall

Auto switching GitHub theme based on vscode theme is a good default, but I'm concerns about automatically switching everyone using dark VS Code themes over to a dark preview without offering a way to switch back to the light preview. Besides matching GitHub, one reason people use this extension is to have the preview be light while the editor remains dark

Let's add a new setting with values: `auto`, `light`, and `dark`. There's not a good story for reading setting values inside the markdown preview, but what you can do is create a very simple [markdown-it extension](https://code.visualstudio.com/api/extension-guides/markdown-extension#adding-support-for-new-syntax-with-markdownit-plugins) that injects the setting value by wrapping the rendered content in something like:


```html
<div class="github-style-dark">
...original content
</div>
```

Then the css just would have to be updated to make sure `github-style-dark` also forces the page to use the dark GitHub theme

**Comment by SNDST00M** [comment]
There should probably also be a `system` setting to use the OS theme (as in `@media (prefers-color-theme: dark)`

**Comment by SNDST00M** [comment]
I've enabled edits by maintainers, would you like to help implement it or mark this with `help-wanted`?

**Comment by SNDST00M** [comment]
Bumping

**Comment by mjbvz** [comment]
I will review PRs but don't have time atm to do lots of work fixing up incomplete ones like this

**Comment by SNDST00M** [comment]
That's okay, I've made a call for help on the extension Slack in the meantime

**Comment by SNDST00M** [comment]
I think it would be better to toggle the classes using a simpler script:

https://code.visualstudio.com/api/extension-guides/markdown-extension#adding-advanced-functionality-with-scripts

**Comment by SNDST00M** [comment]
Does this allow support for posting worker messages like how `image-preview` did? For live switching of course.

**Comment by SNDST00M** [comment]
Works now, what do you think?

**Comment by ProfMoo** [comment]
@SNDST00M Thanks for all the hard work on this. I can't wait to use it.

My 2 cents: Might be worth adding the config value(s) to the README or docs somewhere. Most VSCode extensions include something like
```
"markdown-preview-github-styles.colorMode": [ dark, light, etc.. ]
```
In the README.md in the root level.

**Comment by SNDST00M** [comment]
Done.

**Comment by SNDST00M** [comment]
@mjbvz as I'm sure Moo would concur, this pull looks ready to ship :)

File: package.json [hunk 1]
```diff
@@ -20,25 +20,56 @@
     "url": "https://github.com/mjbvz/vscode-github-markdown-preview-style/issues"
   },
   "engines": {
-    "vscode": "^1.15.0"
+    "vscode": "^1.41.0"
   },
   "categories": [
     "Other"
   ],
+  "activationEvents": [
```
Comment (mjbvz) [inline package.json:28]: These should not be required. VS Code's markdown support will activate this through `markdown.markdownItPlugins`

File: package.json [hunk 2]
```diff
@@ -20,25 +20,56 @@
...
+        }
+      }
     ]
   },
-  "dependencies": {
-    "highlight.js": "^10.7.2"
-  },
   "devDependencies": {
-    "github-markdown-css": "^4.0.0"
+    "eslint": "^7.20.0",
```
Comment (mjbvz) [inline package.json:68]: Remove eslint since it is not related to this change. Would be to see it added in another PR though

File: package.json [hunk 3]
```diff
@@ -20,25 +20,56 @@
...
-      "./node_modules/highlight.js/styles/github.css"
+      "./dist/base.css",
+      "./dist/github-markdown.css",
+      "./dist/highlight-js.css"
+    ],
+    "configuration": [
+      {
+        "title": "Markdown Preview Github Styling",
+        "properties": {
+          "markdown-preview-github-styles.colorMode": {
```
Comment (mjbvz) [inline package.json:46]: [nit] Let's rename this to `Color Theme` instead of `color mode`. That's the term VS Code uses in a few places, such as `workbench.preferredLightColorTheme`

File: dist/github-markdown.css [hunk 4]
```diff
@@ -0,0 +1,1087 @@
+/* Generated from 'node_modules\github-markdown-css\github-markdown.css' */
+
+.markdown-body.markdown-light,
```
Comment (mjbvz) [inline dist/github-markdown.css:3]: Should this be checked in since it is generated? (I know the previous generated file was committed but think was probably a mistake)

File: dist/extension.js [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+    });
+
+    return {
+        extendMarkdownIt(md) {
+            return md.use(exports.plugin);
+        }
+    };
+}
+
+exports.plugin = function(md) {
```
Comment (mjbvz) [inline dist/extension.js:41]: [nit] Does this need to be exported?

File: dist/extension.js [hunk 6]
```diff
@@ -0,0 +1,47 @@
...
+
+const THEME_CONFIGURATION_ID = 'markdown-preview-github-styles';
+const THEME_CONFIGURATION_KEY = 'colorMode';
+
+const THEME_CONFIGURATION_VALUES = {
+    'auto': true,
+    'system': true,
+    'light': true,
+    'dark': true,
+    'high-contrast': true
```
Comment (mjbvz) [inline dist/extension.js:11]: Is `high-contrast` still a valid value here?

File: dist/extension.js [hunk 7]
```diff
@@ -0,0 +1,47 @@
...
+}
+const THEME_CONFIGURATION_AUTO = 'auto';
+
+let settings = vscode.workspace.getConfiguration(THEME_CONFIGURATION_ID, null);
```
Comment (mjbvz) [inline dist/extension.js:15]: [nit] Instead of saving this as a global, just retrieve the setting using the VS Code api when needed. VS Code already caches the settings internally so the call should be fast

**Review by mjbvz — CHANGES_REQUESTED** [review]
Looks like the correct approach overall. Nice work! Let me know if you have any questions about my comments

File: .vscode/launch.json [hunk 8]
```diff
@@ -1,26 +0,0 @@
-// A launch configuration that compiles the extension and then opens it inside a new window
```
[the remark this answers is not in the document]
Comment (SNDST00M) [reply .vscode/launch.json:1]: They are legacy templates and I don't really know where the new versions are

File: dist/extension.js (hunk 6 above)
Comment (SNDST00M) [reply dist/extension.js]: Not yet, I don't think this will be possible without a second PR :(

File: package.json (hunk 3 above)
Comment (SNDST00M) [reply package.json]: Done as of b2c8012. Was thinking about this as I didn't want users to think there was arbitrary themes available à la GitHub.

File: package.json (hunk 2 above)
Comment (SNDST00M) [reply package.json]: Done as of b2c8012

File: package.json [hunk 9]
```diff
@@ -1,6 +1,6 @@
 {
   "name": "markdown-preview-github-styles",
-  "version": "0.2.0",
+  "version": "0.3.0",
```
[the remark this answers is not in the document]
Comment (SNDST00M) [reply package.json]: Done as of b2c8012

File: dist/github-markdown.css (hunk 4 above)
Comment (SNDST00M) [reply dist/github-markdown.css]: Checked out that file in b2c8012 👍🏾

File: package.json (hunk 1 above)
Comment (SNDST00M) [reply package.json]: Done in ebe7d18

File: dist/extension.js (hunk 5 above)
Comment (SNDST00M) [reply dist/extension.js]: Was thinking of test cases but privatised as of b2c8012

File: dist/extension.js (hunk 7 above)
Comment (SNDST00M) [reply dist/extension.js]: Resolved in c4f0515

File: .vscode/launch.json (hunk 8 above)
[the remark this answers is not in the document]
Comment (SNDST00M) [reply .vscode/launch.json:1]: Tried to restore & update them in 92665c3 - please review @mjbvz!

**Comment by mjbvz** [comment]
Great work! Will publish this update shortly

**Comment by mjbvz** [comment]
Published in 1.0.0!

Please give it a quick test pass though. I made a few changes to the implementation and the styling. Specifically I've removed the extra container/padding that was previously around the markdown content and added back a max-width:

Screen Shot 2022-01-07 at 1 58 52 PM

**Comment by SNDST00M** [comment]
I really like it, nicely done :)

**Comment by junaga** [comment]
Way to go guys!! thank you. I love this extension

**Comment by ProfMoo** [comment]
Thanks to all involved :)
